### PR TITLE
[CLEANUP] Remove deprecated auto-location

### PR DESCRIPTION
**Pull request by chriskrycho**
Current status: in progress *and needs someone else to wrap it up*. This gets most of the way there, but needs the last few pieces pulled across the line. cc. @sandstrom – if you have bandwidth to drive this, please say so! I would *like* to have finished it, but work has been 🔥 for this whole quarter.

### CODE REVIEW

**Comment by chriskrycho** [comment]
I have, unfortunately, completely lost the thread on "what is remaining" from when I started on this many weeks ago. 😩

**Comment by chriskrycho** [comment]
@sandstrom I was able to pick this back up and work on it a bit, and hit a thing which is not specified in the RFC: the exported `Location` from `@ember/routing/location` is now *completely empty*—it is literally just an empty object with nothing on it!—so it's not clear to me what (if anything!) you expected to keep that around for. I think it's fine if we drop it and make it a type-only export which just names the interface which implementations have to conform to, but we should update the RFC with that as "learned during implementation" if that makes sense to you as a path forward. (I think that's more or less what we *have* to do here, as the remaining object is completely useless, but I wanted to run it by you first!)

**Comment by sandstrom** [comment]
> the exported Location from @ember/routing/location is now completely empty [...] I think it's fine if we drop it and make it a type-only export which just names the interface which implementations have to conform to

Sounds like a good idea!

***

I've added a few other comments above.

File: packages/@ember/routing/type-tests/location.test.ts [hunk 1]
```diff
@@ -1,24 +1,24 @@
...
-expectTypeOf<ILocation['initState']>().toEqualTypeOf<(() => void) | undefined>();
-expectTypeOf<ILocation['destroy']>().toEqualTypeOf<() => void>();
+expectTypeOf<Location['implementation']>().toEqualTypeOf<string>();
+expectTypeOf<Location['cancelRouterSetup']>().toEqualTypeOf<boolean | undefined>();
+expectTypeOf<Location['getURL']>().toEqualTypeOf<() => string>();
+expectTypeOf<Location['setURL']>().toEqualTypeOf<(url: string) => void>();
+expectTypeOf<Location['replaceURL']>().toEqualTypeOf<((url: string) => void) | undefined>();
+expectTypeOf<Location['onUpdateURL']>().toEqualTypeOf<(callback: (url: string) => void) => void>();
+expectTypeOf<Location['formatURL']>().toEqualTypeOf<(url: string) => string>();
+expectTypeOf<Location['detect']>().toEqualTypeOf<(() => void) | undefined>();
```
[the remark this answers is not in the document]
Comment (chriskrycho) [reply packages/@ember/routing/type-tests/location.test.ts:16]: Ah, I missed that! Excellent.

File: packages/@ember/routing/location.ts [hunk 2]
```diff
@@ -75,56 +59,93 @@ export type UpdateCallback = (url: string) => void;
...
-export default {
+export interface Location {
+  /**
+   * Returns the string name used to reference the implementation.
+   *
+   * @property implementation
+   * @type String
+   * @public
+   */
+  implementation: keyof Registry & string;
```
[the remark this answers is not in the document]
Comment (chriskrycho) [reply packages/@ember/routing/location.ts:75]: Thanks – I'll double check it!

**Pull request merged**